File: backend/app/api/v1/routers/scenarios.py
from fastapi import APIRouter, Depends, Query, Request, Response

from app.api.deps import get_current_user, get_optional_user
from app.api.service_deps import (
    get_scenario_demo_run_service,
    get_scenario_game_service,
    get_scenario_service,
)
from app.api.support.rate_limit import RateLimitRule, enforce_request_rate_limits
from app.infrastructure.db.models import User
from app.modules.scenarios.model.scenario import (
    ScenarioDemoRunStatusRead,
    ScenarioDemoRunTrackRead,
    ScenarioDemoRunTrackWrite,
    ScenarioGameClaimRead,
    ScenarioGameClaimWrite,
    ScenarioGameEarnRead,
    ScenarioGameEarnWrite,
    ScenarioGameStateRead,
    ScenarioHomeAggregateRead,
    ScenarioRunEventRead,
    ScenarioWorkspaceRead,
    ScenarioWorkspaceTrackWrite,
)
from app.modules.scenarios.service.scenario_demo_run_service import ScenarioDemoRunService
from app.modules.scenarios.service.scenario_game_service import ScenarioGameService
from app.modules.scenarios.service.scenario_service import ScenarioService

router = APIRouter(prefix="/scenarios", tags=["scenarios"])

_DEMO_RUN_TRACK_LIMITS: tuple[RateLimitRule, ...] = (
    RateLimitRule(key_template="scenarios:demo-run:track:ip:{ip}", limit=45, window_seconds=60),
    RateLimitRule(key_template="scenarios:demo-run:track:ip:{ip}", limit=300, window_seconds=3600),
)

_GAME_EARN_LIMITS: tuple[RateLimitRule, ...] = (
    RateLimitRule(key_template="scenarios:game:earn:ip:{ip}", limit=60, window_seconds=60),
    RateLimitRule(key_template="scenarios:game:earn:ip:{ip}", limit=400, window_seconds=3600),
)


@router.get("/aggregate", response_model=ScenarioHomeAggregateRead)
async def scenarios_home_aggregate(
    limit: int = Query(default=8, ge=3, le=24),
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioService = Depends(get_scenario_service),
) -> ScenarioHomeAggregateRead:
    return await svc.get_home_aggregate(viewer, limit=limit)


@router.get("/workspace", response_model=ScenarioWorkspaceRead)
async def scenarios_workspace(
    recent_limit: int = Query(default=8, ge=1, le=24),
    saved_limit: int = Query(default=24, ge=1, le=64),
    unfinished_limit: int = Query(default=6, ge=1, le=24),
    current_user: User = Depends(get_current_user),
    svc: ScenarioService = Depends(get_scenario_service),
) -> ScenarioWorkspaceRead:
    return await svc.get_workspace(
        current_user,
        recent_limit=recent_limit,
        saved_limit=saved_limit,
        unfinished_limit=unfinished_limit,
    )


@router.post("/workspace/track", response_model=ScenarioRunEventRead)
async def scenarios_workspace_track(
    body: ScenarioWorkspaceTrackWrite,
    current_user: User = Depends(get_current_user),
    svc: ScenarioService = Depends(get_scenario_service),
) -> ScenarioRunEventRead:
    return await svc.track_workspace_action(current_user, body)


@router.get("/demo-run/status", response_model=ScenarioDemoRunStatusRead)
async def scenarios_demo_run_status(
    request: Request,
    response: Response,
    prompt_slug: str = Query(min_length=1, max_length=200),
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioDemoRunService = Depends(get_scenario_demo_run_service),
) -> ScenarioDemoRunStatusRead:
    return await svc.get_status(
        prompt_slug=prompt_slug,
        viewer=viewer,
        request=request,
        response=response,
    )


@router.post("/demo-run/track", response_model=ScenarioDemoRunTrackRead)
async def scenarios_demo_run_track(
    request: Request,
    response: Response,
    body: ScenarioDemoRunTrackWrite,
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioDemoRunService = Depends(get_scenario_demo_run_service),
) -> ScenarioDemoRunTrackRead:
    await enforce_request_rate_limits(request, _DEMO_RUN_TRACK_LIMITS)
    return await svc.track_run(
        body=body,
        viewer=viewer,
        request=request,
        response=response,
    )


@router.get("/game/state", response_model=ScenarioGameStateRead)
async def scenarios_game_state(
    request: Request,
    response: Response,
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioGameService = Depends(get_scenario_game_service),
) -> ScenarioGameStateRead:
    return await svc.get_state(
        viewer=viewer,
        request=request,
        response=response,
    )


@router.post("/game/earn", response_model=ScenarioGameEarnRead)
async def scenarios_game_earn(
    request: Request,
    response: Response,
    body: ScenarioGameEarnWrite,
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioGameService = Depends(get_scenario_game_service),
) -> ScenarioGameEarnRead:
    await enforce_request_rate_limits(request, _GAME_EARN_LIMITS)
    return await svc.earn(
        body=body,
        viewer=viewer,
        request=request,
        response=response,
    )


@router.post("/game/claim", response_model=ScenarioGameClaimRead)
async def scenarios_game_claim(
    request: Request,
    response: Response,
    body: ScenarioGameClaimWrite,
    viewer: User | None = Depends(get_optional_user),
    svc: ScenarioGameService = Depends(get_scenario_game_service),
) -> ScenarioGameClaimRead:
    return await svc.claim(
        body=body,
        viewer=viewer,
        request=request,
        response=response,
    )

File: frontend/lib/client-api/scenarios.ts
import { API_ENDPOINTS } from "../constants/api";
import type {
  ScenarioDemoRunStatusRead,
  ScenarioDemoRunTrackRead,
  ScenarioDemoRunTrackRequest,
  ScenarioGameClaimRead,
  ScenarioGameClaimRequest,
  ScenarioGameEarnRead,
  ScenarioGameEarnRequest,
  ScenarioGameStateRead,
  ScenarioHomeAggregateRead,
  ScenarioRunEventRead,
  ScenarioWorkspaceRead,
  ScenarioWorkspaceTrackRequest,
} from "../types";
import { authFetch, jsonInit } from "./transport";

export async function fetchScenarioWorkspace(params?: {
  recent_limit?: number;
  saved_limit?: number;
  unfinished_limit?: number;
}): Promise<ScenarioWorkspaceRead> {
  const query = new URLSearchParams();
  if (params?.recent_limit) {
    query.set("recent_limit", String(params.recent_limit));
  }
  if (params?.saved_limit) {
    query.set("saved_limit", String(params.saved_limit));
  }
  if (params?.unfinished_limit) {
    query.set("unfinished_limit", String(params.unfinished_limit));
  }
  const suffix = query.toString();
  const path = suffix ? `${API_ENDPOINTS.scenariosWorkspace}?${suffix}` : API_ENDPOINTS.scenariosWorkspace;
  return authFetch<ScenarioWorkspaceRead>(path);
}

export async function trackScenarioWorkspaceAction(body: ScenarioWorkspaceTrackRequest): Promise<ScenarioRunEventRead> {
  return authFetch<ScenarioRunEventRead>(
    API_ENDPOINTS.scenariosWorkspaceTrack,
    jsonInit("POST", body),
  );
}

export async function fetchScenarioHomeAggregateClient(limit = 8): Promise<ScenarioHomeAggregateRead> {
  const path = `${API_ENDPOINTS.scenariosAggregate}?limit=${encodeURIComponent(String(limit))}`;
  return authFetch<ScenarioHomeAggregateRead>(path);
}

export async function fetchScenarioDemoRunStatus(promptSlug: string): Promise<ScenarioDemoRunStatusRead> {
  const path = `${API_ENDPOINTS.scenariosDemoRunStatus}?prompt_slug=${encodeURIComponent(promptSlug)}`;
  return authFetch<ScenarioDemoRunStatusRead>(path);
}

export async function trackScenarioDemoRun(body: ScenarioDemoRunTrackRequest): Promise<ScenarioDemoRunTrackRead> {
  return authFetch<ScenarioDemoRunTrackRead>(API_ENDPOINTS.scenariosDemoRunTrack, jsonInit("POST", body));
}

export async function fetchScenarioGameState(): Promise<ScenarioGameStateRead> {
  return authFetch<ScenarioGameStateRead>(API_ENDPOINTS.scenariosGameState);
}

export async function earnScenarioGameTokens(body: ScenarioGameEarnRequest): Promise<ScenarioGameEarnRead> {
  return authFetch<ScenarioGameEarnRead>(API_ENDPOINTS.scenariosGameEarn, jsonInit("POST", body));
}

export async function claimScenarioGameTokens(body?: ScenarioGameClaimRequest): Promise<ScenarioGameClaimRead> {
  return authFetch<ScenarioGameClaimRead>(API_ENDPOINTS.scenariosGameClaim, jsonInit("POST", body ?? {}));
}
